src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+                }
+                else
+                {
+                    size *= 2;
+                }
+            }
+            while (result == Marshal.SizeOf<ulong>() * threadIds.Length);
+
+            Debug.Assert((result % Marshal.SizeOf<ulong>()) == 0);
+
+            // Remove any extra elements
+            Array.Resize<ulong>(ref threadIds, (int)(result / Marshal.SizeOf<ulong>()));
+
+            // Loop over each thread and get the thread info
+            List<KeyValuePair<ulong, proc_threadinfo?>> threads = new List<KeyValuePair<ulong, proc_threadinfo?>>();
+            foreach (ulong id in threadIds)
+            {
+                threads.Add(new KeyValuePair<ulong, proc_threadinfo?>(id, GetThreadInfoById(pid, id)));
+            }
+
+            return threads;
+        }
+
+        /// <summary>
+        /// Retrieves the number of open file descriptors for the specified pid
+        /// </summary>
+        /// <returns>Returns an array of open File Descriptors for this process</returns>
+        /// <remarks>
+        /// This function doesn't use the helper since it seems to allow passing NULL
+        /// values in to the buffer and length parameters to get back an estimation 
+        /// of how much data we will need to allocate; the other flavors don't seem
+        /// to support doing that.
+        /// </remarks>
+        internal static unsafe proc_fdinfo[] GetFileDescriptorsForPid(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Query for an estimation about the size of the buffer we will need. This seems
+            // to add some padding from the real number, so we don't need to do that
+            int result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, (proc_fdinfo*)null, 0);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            proc_fdinfo[] fds = null;
+            int size = (int)(result / Marshal.SizeOf<proc_fdinfo>());
+
+            // Just in case the app opened a ton of handles between when we asked and now,
+            // make sure we retry if our buffer is filled
+            do
+            {
+                fds = new proc_fdinfo[size];
+                fixed (proc_fdinfo* pFds = fds)
+                {
+                    result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, pFds, Marshal.SizeOf<proc_fdinfo>() * fds.Length);
+                }
+
+                if (result <= 0)
+                {
+                    throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error());
+                }
+                else
+                {
+                    size *= 2;
+                }
+            }
+            while (result == (fds.Length * Marshal.SizeOf<proc_fdinfo>()));
```

Comment:
This function seems to add a buffer amount to the first call when I pass NULL as the parameters (see my comment to the first call to proc_pidinfo) so that's where the buffer comes in.

Good catch; since I'm not iterating over the array here, I should resize it to the correct amount to trim off any extra elements.

Also adding a checked block around the *= 2